sdks/java/io/mqtt/src/test/java/org/apache/beam/sdk/io/mqtt/MqttIOTest.java

```diff
@@ -0,0 +1,197 @@
+    broker = new BrokerService();
+    broker.setUseJmx(false);
+    broker.setPersistenceAdapter(new MemoryPersistenceAdapter());
+    broker.addConnector(new URI("mqtt://localhost:" + port));
+    broker.start();
+  }
+
+  @Test
+  @Category(NeedsRunner.class)
+  public void testRead() throws Exception {
+    Pipeline pipeline = TestPipeline.create();
+
+    PCollection<byte[]> output = pipeline.apply(
+        MqttIO.read()
+            .withConnectionConfiguration(
+                MqttIO.ConnectionConfiguration.create(
+                    "tcp://localhost:" + port,
+                    "READ_TOPIC"))
+          .withMaxNumRecords(10));
+
+    PAssert.thatSingleton(output.apply("Count", Count.<byte[]>globally()))
+        .isEqualTo(10L);
+    PAssert.that(output).satisfies(new SerializableFunction<Iterable<byte[]>, Void>() {
+      @Override
+      public Void apply(Iterable<byte[]> input) {
+        for (byte[] element : input) {
+          String inputString = new String(element);
+          Assert.assertTrue(inputString.startsWith("This is test "));
+          int count = Integer.parseInt(inputString.substring("This is test ".length()));
+          Assert.assertTrue(count < 10);
+          Assert.assertTrue(count >= 0);
+        }
+        return null;
+      }
+    });
+
+    // produce messages on the broker in another thread
+    // This thread prevents to block the pipeline waiting for new messages
+    Thread thread = new Thread() {
+      public void run() {
+        try {
+          // gives time to the pipeline to start
+          Thread.sleep(2000);
+        } catch (Exception e) {
+          // nothing to do
+        }
+        try {
+          MqttClient client = new MqttClient("tcp://localhost:" + port,
+              MqttClient.generateClientId());
+          client.connect();
+          for (int i = 0; i < 10; i++) {
+            MqttMessage message = new MqttMessage();
+            message.setQos(0);
+            message.setPayload(("This is test " + i).getBytes());
+            client.publish("READ_TOPIC", message);
+          }
+          client.disconnect();
+          client.close();
+        } catch (Exception e) {
+          // nothing to do
+        }
+      }
+    };
+    thread.start();
+
+    pipeline.run();
+  }
+
+  @Test
+  @Category(NeedsRunner.class)
+  public void testWrite() throws Exception {
+    List<MqttMessage> messages = new ArrayList<>();
+    MqttClient client = receive(messages);
+
+    Pipeline pipeline = TestPipeline.create();
+
+    ArrayList<byte[]> data = new ArrayList<>();
+    for (int i = 0; i < 100; i++) {
+      data.add("Test".getBytes());
+    }
+    pipeline.apply(Create.of(data))
+        .apply(MqttIO.write()
+            .withConnectionConfiguration(
+                MqttIO.ConnectionConfiguration.create(
+                    "tcp://localhost:" + port,
+                    "WRITE_TOPIC"))
+        .withQoS(1));
+    pipeline.run();
+
+    Assert.assertEquals(100, messages.size());
```

Comment:
Again, this is incorrect. After pipeline.run() has completed, messages may be buffered inside the broker, but it has not necessarily delivered them to the client, nor has the client necessarily already added them to `messages`.